lib/ts/test_Ptr.cc

```diff
@@ -95,5 +95,5 @@ REGRESSION_TEST(Ptr_bool)(RegressionTest *t, int /* atype ATS_UNUSED */, int *ps
   Ptr<PtrObject> some = make_ptr(new PtrObject(&alive));
 
   box.check(!none, "Empty Ptr<T> is false");
-  box.check(some, "Non-empty Ptr<T> is true");
+  box.check(some != nullptr, "Non-empty Ptr<T> is true");
```

Comment:
This case is supposed to test the (implicit) conversion to `bool`, not the `!=` operator.

Reply:
But implicit conversion to bool, other than in if statements or as arguments to logical operators, causes problems.  If you look at the link mentioned in the comments, I am just making the change to use the solution to this problem added to the base language in C++11.

Reply:
AFAIK implicit bool conversion was OK prior to this PR, and even after the change to the C++11 technique, seems to have worked for compilers other than ICC. If we believe that we should never have done the implicit conversion, that's fine, but in that case we should update the tests and commit message appropriately.

For example, maybe we should be testing
```C
box.check(some == true, "Non-empty Ptr<T> is true");
box.check(some != false, "Non-empty Ptr<T> is true");
```

Reply:
I personally don't feel like there is any good choice with this.  The complex rules for implicit conversion are one of the many ugly corner cases that turn people off from C++.  My thinking is, between two bad options, it probably will turn out the least bad to pick the standardized bad option.  Particularly if at some point we want to move from Ptr to std::shared_ptr .

Reply:
'some == true' won't fly:

  CXX    test_Ptr.o
test_Ptr.cc: In function ‘void RegressionTest_Ptr_bool(RegressionTest*, int, int*)’:
test_Ptr.cc:98:53: error: no match for ‘operator==’ (operand types are ‘Ptr<PtrObject>’ and ‘bool’)
   box.check(some and (bool(some) == true) and (some == true), "Non-empty Ptr<T> is true");

Reply:
The point @jpeach is that the compiler will invoke different code for operators vs implicit conversion.  We can agree that you should say if (x == value) but the issue here is that a operator == will be called, not the implicit operator. The original test was set to invoke the correct operator overload.  

I agree with jpeach in that we should avoid implicit conversions.. and it would be better for me to have overloaded operator == and != in this class to force that all code is in the form of:

if(x==y)  no in the form of if(x)

Reply:
I don't want to  an equality operator for `bool` just to make testing code a bit more compact. I think the code will need to do something like
```
if (x) REQUIRE(true, "true");
else REQUIRE(false, "Explicit bool conversion didn't work");
```
I'm not sure a ternary operator will always be as good a choice, e.g.
```
REQUIRE(x ? true : false, "Explicit bool conversion to true");
```